src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -298,10 +298,10 @@ internal ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken)
             {
                 CommitUnsynchronized();
 
-                _readerAwaitable.Complete(out completionData);
-
                 cancellationTokenRegistration = _writerAwaitable.AttachToken(cancellationToken, s_signalWriterAwaitable, this);
 
+                _readerAwaitable.Complete(out completionData);
```

Comment:
Can you please add a comment stating why this ordering is necessary for correctness?